### ISSUE TITLE
Subdirectory installs: API urls are not generated correctly 

### ISSUE THREAD

**Issue by thompsonsj** [cloudcreativity/laravel-json-api#202]
If the Laravel application is installed in a subdirectory, URLs (e.g. self and related) are generated relative to the domain rather than the application URL.

Consider an API being served from `https://example.org/app/` with the following default configuration in `config/json-api-default.php`.

```
  'url' => [
        'host' => null,
        'namespace' => '/api/v1',
        'name' => 'api:v1:',
    ],
```

API results look like:

```
HTTP/1.1 200 OK
Content-Type application/vnd.api+json

{
    "data": [
        {
            "type": "posts",
            "id": "1",
            "attributes": {
                "title": "My First Post",
                "slug": "my-first-post"
            },
            "links": {
                "self": "https://example.org/api/v1/posts/1"
            }
        },
        {
            "type": "posts",
            "id": "2",
            "attributes": {
                "title": "Hello World",
                "slug": "hello-world"
            },
            "links": {
                "self": "https://example.org/api/v1/posts/2"
            }
        }
    ]
}
```

The links should respect the application URL and be of the form `https://example.org/app/api/v1/posts/2`.

It seems that this issue may be caused by the use of `getSchemeAndHttpHost()` from the Symfony HttpFoundation package in `Http\Middleware\BootJsonApi.php` (lines 77 and 78).

```
/** Build and register the API */
$api = $this->bindApi($namespace, $request->getSchemeAndHttpHost());
```

If we set the second argument to `null`, urls are generated correctly.

```
/** Build and register the API */
$api = $this->bindApi($namespace, null);
```

**Comment by lindyhopchris** [comment]
Hi! Surely your `namespace` setting should be `/app/api/v1` because that is your API's namespace on the host?

**Comment by lindyhopchris** [comment]
Or alternatively, what Laravel settings are you using to tell Laravel that it is hosted under the `app/` folder?

**Comment by lindyhopchris** [comment]
Would seem that this needs to be supported, but we cannot removed the detection of the host because otherwise it will break the host working correctly on wildcard sub-domains. Changing that line to pass a second argument as `null` would break that use case.

**Comment by thompsonsj** [comment]
Yes, absolutely. setting `namespace` to `/app/api/v1` solves the problem. However, because `config/json-api-default.php` is committed to version control for a given Laravel app, it makes it awkward to manage installations of the same application in different directories. This might be a less common use case though!

**Comment by thompsonsj** [comment]
Using an ENV setting is an excellent idea, thank you! I can't believe I didn't think of that.

I am going do some testing just now as I note that `CloudCreativity\LaravelJsonApi\Api\Repository` is called twice upon visiting an endpoint. Once with no `$host` value (`CloudCreativity\LaravelJsonApi\Routing\ResourceRegistrar::api`) and then once with a host value (`CloudCreativity\LaravelJsonApi\Http\Middleware\BootJsonApi::createApi`).

As for normal routing, I am using the following.

```
Route::get('/home', 'HomeController@home');
```

**Comment by thompsonsj** [comment]
Using an ENV setting caused routing errors.

I propose an alternative solution combining Symfony's `getBaseUrl` with `getSchemeAndHttpHost`:

https://github.com/cloudcreativity/laravel-json-api/pull/204

**Comment by lindyhopchris** [comment]
Thanks for investigating this. Can't see how to add tests for it unfortunately so have just merged it for the moment as it doesn't break any existing stuff as far as I can see.

In terms of `createApi` being called twice. In development that will be once for the JSON API routes being registered and once for the actual request being processed. In production it would only be called once if you've used route caching.

I'm planning to release this change and another as `v1.0.0-alpha.4` but need to hear back about another issue before tagging. If you want to use in the meantime, then use the following:

```bash
composer require cloudcreativity/laravel-json-api:dev-master
```

I don't make breaking changes to `master` so it's safe to use until I tag.

**Comment by thompsonsj** [comment]
Thank you for resolving this so quickly and merging this in.
